# 04 · Modelado — escenario histórico de todos los siniestros
### Fase 4 de CRISP-DM — OE2, OE3

> **Corrección metodológica · 01/10/2026:** la validación aleatoria de la sección 10
> queda **ANULADA**. Se retiraron sus llamadas ejecutables y sus salidas, no solo
> se cambió su nombre. Sus métricas no sustentan selección de modelos ni
> generalización temporal. La sección 10 identifica la evidencia temporal de 05.

Este notebook conserva el experimento original con todos los siniestros, incluidos
los de solo daños. **No representa el modelo principal ni alimenta el dashboard.**
La línea principal usa víctimas y se documenta en 04B → 04C → 05B; su autoridad es
[`modelo_principal.json`](../models/victimas/modelo_principal.json).

**Objetivo histórico:** entrenar y comparar Regresión Logística, Random Forest
y XGBoost para la etiqueta `Alto_Riesgo` del escenario total.

El split final sigue siendo entrenamiento 2018–2022 y evaluación retrospectiva
2023–2024. El umbral de etiqueta de ese split se calculó con 2018–2022.
**Esto no basta para validar cortes internos:** en cada corte también deben
recalcularse la etiqueta, el preprocesamiento y el modelo usando solo su pasado.

Las salidas conservadas de entrenamiento y comparación son históricas; esta
corrección no las reejecutó ni alteró artefactos. No deben presentarse como una
nueva ejecución. Ejecutar todo el notebook vuelve a entrenar y su sección 12
escribe modelos: hacerlo únicamente en una copia aislada, no sobre el cierre.

Estructura:
1. Introducción y alcance histórico
2. Importación de librerías
3. Carga del dataset procesado
4. Definición de variables X e y
5. Split final temporal
6. Preprocesamiento
7. Regresión Logística
8. Random Forest
9. XGBoost
10. Anulación de validación aleatoria y referencia temporal
11. Comparación retrospectiva histórica
12. Guardado histórico de modelos

Corrección y procedencia: [acta de anulación](../reports/correcciones/04_validacion_aleatoria_anulada.md).


## 2. Importación de librerías

In [ ]:
import pandas as pd
import numpy as np
import joblib
from pathlib import Path

from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score, roc_auc_score, precision_score, recall_score

import xgboost as xgb

pd.set_option('display.max_columns', 50)

PROCESSED_PATH = Path('../data/processed')
MODELS_PATH = Path('../models')
MODELS_PATH.mkdir(parents=True, exist_ok=True)

RANDOM_STATE = 42


## 3. Carga del dataset procesado

In [2]:
dataset = pd.read_parquet(PROCESSED_PATH / 'dataset_localidad_franja_fecha.parquet')
print('Filas:', dataset.shape[0], '| Columnas:', dataset.shape[1])
dataset.head()


Filas: 204560 | Columnas: 16


,Localidad,Franja_Horaria,Fecha_Acc,Num_Accidentes,Dia_Semana,Mes,Es_Fin_de_Semana,Es_Festivo,Accidentes_Prom_7d,Accidentes_Prom_30d,Accidentes_Semana_Anterior,Sin_Historial_Accidentes_Prom_7d,Sin_Historial_Accidentes_Prom_30d,Sin_Historial_Accidentes_Semana_Anterior,Periodo,Alto_Riesgo
0,ANTONIO NARIÑO,Madrugada,2018-01-01,0,Monday,1,0,1,0.0,0.0,0.0,1,1,1,train,0
1,ANTONIO NARIÑO,Madrugada,2018-01-02,0,Tuesday,1,0,0,0.0,0.0,0.0,0,0,1,train,0
2,ANTONIO NARIÑO,Madrugada,2018-01-03,0,Wednesday,1,0,0,0.0,0.0,0.0,0,0,1,train,0
3,ANTONIO NARIÑO,Madrugada,2018-01-04,0,Thursday,1,0,0,0.0,0.0,0.0,0,0,1,train,0
4,ANTONIO NARIÑO,Madrugada,2018-01-05,0,Friday,1,0,0,0.0,0.0,0.0,0,0,1,train,0


## 4. Definición de variables X e y

**Variables excluidas de `X` y por qué:**

| Columna | Motivo de exclusión |
|---|---|
| `Fecha_Acc` | Identificador temporal, no generaliza; ya está representada por `Dia_Semana`, `Mes`, `Es_Fin_de_Semana`, `Es_Festivo` |
| `Num_Accidentes` | Es la misma información usada para construir `Alto_Riesgo` — incluirla sería *data leakage* |
| `Periodo` | Se usa para el split, no como predictora |

Todo lo demás (calendario + históricas + flags de historial insuficiente) sí es válido como predictor.


In [3]:
cat_cols = ['Localidad', 'Franja_Horaria', 'Dia_Semana']

num_cols = [
    'Mes', 'Es_Fin_de_Semana', 'Es_Festivo',
    'Accidentes_Prom_7d', 'Accidentes_Prom_30d', 'Accidentes_Semana_Anterior',
    'Sin_Historial_Accidentes_Prom_7d', 'Sin_Historial_Accidentes_Prom_30d',
    'Sin_Historial_Accidentes_Semana_Anterior'
]

X = dataset[cat_cols + num_cols]
y = dataset['Alto_Riesgo']

print('Variables categóricas:', cat_cols)
print('Variables numéricas:', num_cols)
print('Total de predictoras:', len(cat_cols) + len(num_cols))


Variables categóricas: ['Localidad', 'Franja_Horaria', 'Dia_Semana']
Variables numéricas: ['Mes', 'Es_Fin_de_Semana', 'Es_Festivo', 'Accidentes_Prom_7d', 'Accidentes_Prom_30d', 'Accidentes_Semana_Anterior', 'Sin_Historial_Accidentes_Prom_7d', 'Sin_Historial_Accidentes_Prom_30d', 'Sin_Historial_Accidentes_Semana_Anterior']
Total de predictoras: 12


## 5. Reutilización del split temporal (`Periodo`)

Se reutiliza exactamente el split construido en el Notebook 3 (`train` ≤ 2022-12-31,
`test` = 2023–2024) — no se genera un `train_test_split` nuevo, para no romper la coherencia con
el umbral de `Alto_Riesgo` que ya fue calculado respetando esa misma frontera temporal.


### 5.1 Conjunto de entrenamiento

In [4]:
train_mask = dataset['Periodo'] == 'train'

X_train = X[train_mask]
y_train = y[train_mask]

print('X_train:', X_train.shape)
print('y_train:', y_train.shape)


X_train: (146080, 12)
y_train: (146080,)


### 5.2 Conjunto de prueba

In [5]:
test_mask = dataset['Periodo'] == 'test'

X_test = X[test_mask]
y_test = y[test_mask]

print('X_test:', X_test.shape)
print('y_test:', y_test.shape)


X_test: (58480, 12)
y_test: (58480,)


### 5.3 Distribución de la variable objetivo

In [6]:
resumen_dist = pd.DataFrame({
    'train': y_train.value_counts(normalize=True).sort_index(),
    'test': y_test.value_counts(normalize=True).sort_index()
})
resumen_dist.index = ['No Alto Riesgo (0)', 'Alto Riesgo (1)']
resumen_dist.round(4)


,train,test
No Alto Riesgo (0),0.8207,0.9369
Alto Riesgo (1),0.1793,0.0631


### Interpretación

La proporción de `Alto_Riesgo` es notablemente más baja en test (2023–2024) que en train
(2018–2022). Esto ya se identificó en el Notebook 3 y se relaciona con el menor volumen de
registros de 2023–2024 respecto a años anteriores (posible subregistro), no con una reducción
real de accidentes. Este desbalance entre periodos se debe tener presente al interpretar las
métricas del modelo sobre test en el Notebook 05 — un desempeño más bajo ahí no necesariamente
indica que el modelo generaliza mal, sino que el período de prueba tiene una distribución distinta
al de entrenamiento (*data drift*).


## 6. Preprocesamiento

**Regla general para evitar fuga de información:** todo transformador (encoder, scaler) se ajusta
(`fit`) **solo con `X_train`**, y luego se aplica (`transform`) tanto a train como a test. Ajustar
con el dataset completo filtraría información del período de prueba hacia el entrenamiento.


### 6.1 Codificación de variables categóricas

`Localidad`, `Franja_Horaria` y `Dia_Semana` se codifican con One-Hot Encoding — ninguno de los
tres modelos acepta texto directamente. `handle_unknown='ignore'` evita errores si test tuviera
alguna categoría no vista en train (no debería ocurrir aquí, pero es buena práctica).


In [7]:
encoder_cat = OneHotEncoder(handle_unknown='ignore')
encoder_cat.fit(X_train[cat_cols])

print('Categorías codificadas por variable:')
for col, cats in zip(cat_cols, encoder_cat.categories_):
    print(f'  {col}: {len(cats)} categorías')


Categorías codificadas por variable:
  Localidad: 20 categorías
  Franja_Horaria: 4 categorías
  Dia_Semana: 7 categorías


### 6.2 Escalado de variables numéricas

**Solo necesario para Regresión Logística** — es sensible a la escala de las variables (`Mes` va
de 1 a 12, `Accidentes_Prom_30d` puede ir de 0 a 14). Random Forest y XGBoost son invariantes a
escala, así que para esos dos se usan las variables numéricas sin transformar.


In [8]:
# Preprocesador para Regresión Logística: One-Hot + Escalado
preprocesador_logistica = ColumnTransformer([
    ('cat', OneHotEncoder(handle_unknown='ignore'), cat_cols),
    ('num', StandardScaler(), num_cols)
])

# Preprocesador para modelos de árboles: solo One-Hot, numéricas sin transformar
preprocesador_arboles = ColumnTransformer([
    ('cat', OneHotEncoder(handle_unknown='ignore'), cat_cols)
], remainder='passthrough')

preprocesador_logistica.fit(X_train)
preprocesador_arboles.fit(X_train)

X_train_log = preprocesador_logistica.transform(X_train)
X_test_log = preprocesador_logistica.transform(X_test)

X_train_arb = preprocesador_arboles.transform(X_train)
X_test_arb = preprocesador_arboles.transform(X_test)

print('Shape tras preprocesamiento (Logística):', X_train_log.shape)
print('Shape tras preprocesamiento (Árboles):', X_train_arb.shape)


Shape tras preprocesamiento (Logística): (146080, 40)
Shape tras preprocesamiento (Árboles): (146080, 40)


### 6.3 Manejo del desbalance de clases

Con ~18% de casos positivos en train, los tres modelos usan su mecanismo nativo de balanceo en
vez de remuestrear los datos (SMOTE, undersampling), para no alterar artificialmente la
distribución temporal ya construida:

- **Regresión Logística / Random Forest:** `class_weight='balanced'` — pondera automáticamente el
  error de la clase minoritaria en proporción inversa a su frecuencia.
- **XGBoost:** `scale_pos_weight` — se calcula manualmente como la razón negativos/positivos en train.


In [9]:
scale_pos_weight = (y_train == 0).sum() / (y_train == 1).sum()
print(f'scale_pos_weight para XGBoost: {scale_pos_weight:.3f}')


scale_pos_weight para XGBoost: 4.576


## 7. Modelo 1: Regresión Logística

Modelo baseline — simple e interpretable, sirve como punto de referencia para evaluar si los
modelos más complejos (Random Forest, XGBoost) realmente aportan una mejora.


In [10]:
modelo_logistica = LogisticRegression(
    class_weight='balanced',
    max_iter=1000,
    random_state=RANDOM_STATE
)

modelo_logistica.fit(X_train_log, y_train)

pred_logistica = modelo_logistica.predict(X_test_log)
proba_logistica = modelo_logistica.predict_proba(X_test_log)[:, 1]

print('Regresión Logística — desempeño en test:')
print(f'  F1-score:  {f1_score(y_test, pred_logistica):.4f}')
print(f'  AUC-ROC:   {roc_auc_score(y_test, proba_logistica):.4f}')
print(f'  Precision: {precision_score(y_test, pred_logistica):.4f}')
print(f'  Recall:    {recall_score(y_test, pred_logistica):.4f}')


Regresión Logística — desempeño en test:
  F1-score:  0.1803
  AUC-ROC:   0.6577
  Precision: 0.1330
  Recall:    0.2798


## 8. Modelo 2: Random Forest

**Nota sobre regularización:** se limita `max_depth=12` y `min_samples_leaf=20`. Sin estas
restricciones, los árboles crecen sin control con este volumen de datos (146,080 filas de train),
generando un modelo de más de 1 GB en disco y con alto riesgo de sobreajuste. Limitar la
profundidad y el tamaño mínimo de hoja reduce el modelo a ~15 MB sin pérdida relevante de
desempeño, y de paso mejora su capacidad de generalización.


In [11]:
modelo_rf = RandomForestClassifier(
    n_estimators=200,
    max_depth=12,
    min_samples_leaf=20,
    class_weight='balanced',
    random_state=RANDOM_STATE,
    n_jobs=-1
)

modelo_rf.fit(X_train_arb, y_train)

pred_rf = modelo_rf.predict(X_test_arb)
proba_rf = modelo_rf.predict_proba(X_test_arb)[:, 1]

print('Random Forest — desempeño en test:')
print(f'  F1-score:  {f1_score(y_test, pred_rf):.4f}')
print(f'  AUC-ROC:   {roc_auc_score(y_test, proba_rf):.4f}')
print(f'  Precision: {precision_score(y_test, pred_rf):.4f}')
print(f'  Recall:    {recall_score(y_test, pred_rf):.4f}')


Random Forest — desempeño en test:
  F1-score:  0.1551
  AUC-ROC:   0.6506
  Precision: 0.1071
  Recall:    0.2809


## 9. Modelo 3: XGBoost

In [12]:
modelo_xgb = xgb.XGBClassifier(
    n_estimators=200,
    scale_pos_weight=scale_pos_weight,
    random_state=RANDOM_STATE,
    eval_metric='logloss'
)

modelo_xgb.fit(X_train_arb, y_train)

pred_xgb = modelo_xgb.predict(X_test_arb)
proba_xgb = modelo_xgb.predict_proba(X_test_arb)[:, 1]

print('XGBoost — desempeño en test:')
print(f'  F1-score:  {f1_score(y_test, pred_xgb):.4f}')
print(f'  AUC-ROC:   {roc_auc_score(y_test, proba_xgb):.4f}')
print(f'  Precision: {precision_score(y_test, pred_xgb):.4f}')
print(f'  Recall:    {recall_score(y_test, pred_xgb):.4f}')


XGBoost — desempeño en test:
  F1-score:  0.1973
  AUC-ROC:   0.7053
  Precision: 0.1476
  Recall:    0.2972


## 10. Validación aleatoria ANULADA

La versión original utilizó `StratifiedKFold(n_splits=5, shuffle=True)` sobre
2018–2022. **Se anulan sus métricas y conclusiones como evidencia de validación
temporal.** Sus llamadas ejecutables y la tabla de resultados fueron retiradas
de este notebook; la versión histórica permanece recuperable en Git.

Esta corrección no sustituye números por resultados inventados ni declara un
reentrenamiento. La validación temporal ya existente está en la sección 20 de
[05_Evaluacion.ipynb](05_Evaluacion.ipynb).


### 10.1 Motivos de la anulación

- Las particiones aleatorias mezclaban fechas: podían entrenar con observaciones
  posteriores a las usadas para validar.
- Los preprocesadores se habían ajustado a todo 2018–2022 antes de dividirlo,
  por lo que cada partición de validación influía en su transformación.
- La etiqueta fija de 2018–2022 no se recalculaba con el pasado de cada corte.
  No es una etiqueta apropiada para simular validaciones en 2020 o 2021.
- La estratificación conserva proporciones de clases, pero no corrige ninguno
  de esos problemas temporales.

El ajuste con todo `X_train` sigue siendo apropiado para el split final
2018–2022 → 2023–2024; el problema descrito corresponde a la validación interna.


### 10.2 Estado de los resultados retirados

Las medias y desviaciones de los cinco folds aleatorios quedan **sin validez
para justificar estabilidad temporal o seleccionar el modelo**. No se promedian
ni se comparan como equivalentes con los resultados de cortes expansivos.

El acta enlazada en la portada identifica el commit anterior, las celdas
originales y la razón de retirada. No queda código ejecutable de esa validación
aleatoria en 04.


### 10.3 Protocolo temporal de referencia

| Entrenamiento | Validación |
|---|---|
| 2018–2019 | 2020 |
| 2018–2020 | 2021 |
| 2018–2021 | 2022 |

La sección 20 de 05 implementa estos cortes para el **escenario total**:
calcula el cuantil 2/3 de localidad–franja con el entrenamiento de cada corte,
aplica esa etiqueta al año validado y ajusta un pipeline nuevo con ese pasado.
También recalcula `scale_pos_weight` para XGBoost por corte.

Los tres históricos excluyen el día objetivo. La evaluación es secuencial:
presupone disponibles los conteos de días anteriores; no simula predecir de
una sola vez todo el año sin observar sus días previos.


### 10.4 Evidencias temporales existentes (no regeneradas en esta corrección)

- [Resultados por corte y modelo](../reports/evaluation/validacion_temporal_2018_2022_detalle.csv).
- [Resumen de los tres cortes](../reports/evaluation/validacion_temporal_2018_2022_resumen.csv).
- [Predicciones fuera de muestra 2020–2022](../reports/evaluation/predicciones_validacion_temporal_2020_2022.parquet).
- [Comparación gráfica temporal](../reports/evaluation/10_validacion_temporal_2018_2022.png).

Estos archivos proceden del experimento temporal conservado en 05, no de los
cinco folds anulados. No son resultados del escenario principal con víctimas,
que se consulta en 04B, 04C y 05B.


### 10.5 Límites de interpretación

La variación entre tres años describe sensibilidad temporal; no es una garantía
de generalización ni una validación prospectiva. El escenario total está
afectado por cambios en los registros de solo daños.

2023–2024 ya se utilizó para calcular y comparar resultados en las secciones
7–9 y 11. **No es un holdout intacto**. Su función es evaluación retrospectiva,
no selección independiente ni confirmación prospectiva. Corregir la validación
aleatoria no revierte la inspección previa de ese periodo.


## 11. Comparación retrospectiva histórica de modelos

Se conservan las métricas calculadas sobre 2023–2024 para los tres modelos del
escenario total. Este periodo ya fue inspeccionado durante el desarrollo:
la tabla no es validación interna ni una evaluación confirmatoria independiente.

Estas salidas históricas no eligen el modelo vigente. El análisis detallado de
este escenario está en 05; la selección del modelo principal con víctimas se
documenta en 04C y su evaluación retrospectiva en 05B.


In [15]:
comparacion = pd.DataFrame({
    'Regresión Logística': {
        'F1-score': f1_score(y_test, pred_logistica),
        'AUC-ROC': roc_auc_score(y_test, proba_logistica),
        'Precision': precision_score(y_test, pred_logistica),
        'Recall': recall_score(y_test, pred_logistica),
    },
    'Random Forest': {
        'F1-score': f1_score(y_test, pred_rf),
        'AUC-ROC': roc_auc_score(y_test, proba_rf),
        'Precision': precision_score(y_test, pred_rf),
        'Recall': recall_score(y_test, pred_rf),
    },
    'XGBoost': {
        'F1-score': f1_score(y_test, pred_xgb),
        'AUC-ROC': roc_auc_score(y_test, proba_xgb),
        'Precision': precision_score(y_test, pred_xgb),
        'Recall': recall_score(y_test, pred_xgb),
    },
}).T.round(4)

comparacion


,F1-score,AUC-ROC,Precision,Recall
Regresión Logística,0.1803,0.6577,0.1330,0.2798
Random Forest,0.1551,0.6506,0.1071,0.2809
XGBoost,0.1973,0.7053,0.1476,0.2972


### Interpretación

La tabla describe el desempeño histórico del escenario total sobre 2023–2024.
La distribución de la etiqueta cambia entre entrenamiento y evaluación, lo que
limita la comparabilidad y exige examinar cobertura y composición de registros.
Estas métricas no prueban causalidad, utilidad operativa ni generalización futura.

No deben combinarse con las métricas de la validación aleatoria anulada ni
confundirse con los resultados del RF ajustado del escenario con víctimas.


## 12. Guardado de modelos

Se guardan los tres modelos entrenados junto con sus respectivos preprocesadores (necesarios para
transformar cualquier dato nuevo de la misma forma antes de predecir), listos para la evaluación
detallada del Notebook 05.


In [16]:
joblib.dump(modelo_logistica, MODELS_PATH / 'modelo_logistica.pkl')
joblib.dump(preprocesador_logistica, MODELS_PATH / 'preprocesador_logistica.pkl')

joblib.dump(modelo_rf, MODELS_PATH / 'modelo_random_forest.pkl')
joblib.dump(modelo_xgb, MODELS_PATH / 'modelo_xgboost.pkl')
joblib.dump(preprocesador_arboles, MODELS_PATH / 'preprocesador_arboles.pkl')

print('Modelos y preprocesadores guardados en:', MODELS_PATH.resolve())
print(sorted(p.name for p in MODELS_PATH.glob('*.pkl')))


Modelos y preprocesadores guardados en: /Users/camilo/Library/CloudStorage/OneDrive-FundaciónUniversitariaKonradLorenz/Siniestralidad_Vial/models
['modelo_logistica.pkl', 'modelo_random_forest.pkl', 'modelo_xgboost.pkl', 'preprocesador_arboles.pkl', 'preprocesador_logistica.pkl']


## Resumen de decisiones y correcciones

| Decisión | Alcance |
|---|---|
| Reutilizar `Periodo` del Notebook 3 | Split final 2018–2022 → 2023–2024 |
| Excluir `Fecha_Acc`, `Num_Accidentes` y `Periodo` de X | Evitar usar el conteo objetivo e identificadores del split como predictores |
| Ajustar transformadores solo con train final | Válido para ese split; no justifica reutilizarlos en la validación interna |
| Escalado para Regresión Logística y ponderación de clases | Configuración del experimento histórico |
| **Anular validación aleatoria** | Código y salidas retirados; no aporta evidencia de estabilidad temporal |
| Referencia temporal: sección 20 de 05 | Tres cortes expansivos con etiqueta, preprocesador y modelo reajustados por corte |
| 2023–2024 es evaluación retrospectiva | No se presenta como test intacto ni prospectivo |
| Preservar artefactos originales | Esta corrección documental no reentrena ni sobrescribe modelos |

**Continuación del escenario histórico:** [05_Evaluacion.ipynb](05_Evaluacion.ipynb).
**Línea principal:** 04B → 04C → 05B, escenario con víctimas.
